refactor(Make all links use a human readable): In posting Journal

lomamech opened the pull request:
- Implement component bhTransactionType
- Implement test E2E for the component bhTransactionType
- Update journal search Modal and display readable link

this issue reference #2139 and #2130

jniles commented on client/src/js/components/bhTransactionType.js:
```diff
@@ -0,0 +1,50 @@
+angular.module('bhima.components')
```
There is already a [bhTransactionTypeSelect](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/js/components/bhTransactionTypeSelect.js).  Can you use that instead?

lomamech replied:
The reason I had created another component and that in the old bhTransactionTypeSelect. we had the possibility of choosing several type transactions whereas in the bhTransactionType we have the option of choosing one

jniles reviewed (commented):
It looks like many of the comments from https://github.com/IMA-WorldHealth/bhima-2.X/pull/2178 would be applicable here.  Let me know if you can adapt the comments into something nice.  If not, I'll be happy to work with you in person on it!

lomamech committed:
refactor(Posting Journal)
- Remove component bhTransactionType
- And Use component bh-transaction-type-select
- Implement component bh-project-select
- Use formatView().customFilters for the lastViewFilters
- Update API journal use custom filters for the reasearch of multiple
  transaction type

lomamech committed:
refactor(Posting Journal)
- Remove the test E2E for the component bhTransactionTypes
- Set the test E2E for the component bhProjectSelect
- Update the research for component Account

jniles commented on client/src/modules/journal/modals/search.modal.js:
```diff
@@ -80,8 +58,18 @@ function JournalSearchModalController(Instance, Notify,
   // custom filter account_id - assign the value to the searchQueries object
   vm.onSelectAccount = function onSelectAccount(account) {
     vm.searchQueries.account_id = account.id;
+    displayValues.account_id = String(account.number).concat(' - ', account.label);
```
There is no need to use `String` - what you are looking for is:
```js
displayValues.account_id = account.hrLabel;
```

The `AccountService` formats labels after they are downloaded - see [this line](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/modules/accounts/accounts.service.js#L45).

jniles commented on client/src/js/components/bhProjectSelect.js:
```diff
@@ -0,0 +1,47 @@
+
+  $ctrl.$onInit = function onInit() {    
+    // fired when a project has been selected
+    $ctrl.onSelectCallback = $ctrl.onSelectCallback || angular.noop;
```
You don't actually need this - angular will automatically assign a `noop()` to it since it so not required (`<?`).

jniles commented on client/src/js/components/bhProjectSelect.js:
```diff
@@ -0,0 +1,47 @@
+      })
+      .catch(Notify.handleError);
+
+    $ctrl.valid = true;
```
Where did `valid` come from?

jniles commented on client/src/js/components/bhProjectSelect.js:
```diff
@@ -0,0 +1,47 @@
+    $ctrl.onSelectCallback = $ctrl.onSelectCallback || angular.noop;
+
+    // default for form name
+    $ctrl.name = $ctrl.name || 'FORM.LABELS.PROJECT';
```
There's no need for this child form.  We can remove it.

jniles commented on client/src/js/components/bhProjectSelect.js:
```diff
@@ -0,0 +1,47 @@
+ * Project Select component
+ *
+ */
+function projectSelectController(Projects, Notify) {
```
Controllers should begin with a capital letter.

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -138,9 +144,10 @@ function buildTransactionQuery(options, posted) {
-  filters.equals('origin_id');
   filters.equals('record_uuid');
 
+  filters.custom('origin_id', 'p.origin_id IN (?)', [typeIds]);
```
:+1:

jniles commented on client/src/modules/templates/bhProjectSelect.tmpl.html:
```diff
@@ -0,0 +1,23 @@
+<div ng-form="{{ $ctrl.name }}" bh-project-select ng-model-options="{ updateOn: 'default' }">
```
You can remove this `ng-form` call.  it doesn't actually do anything.

jniles commented on client/src/js/components/bhProjectSelect.js:
```diff
@@ -0,0 +1,47 @@
+      disable          : '<?',
+      onSelectCallback : '&',
+      name             : '@?',
+      required         : '<?',      
```
Go ahead and remove `required` and `disable`.  You never use them.

jniles commented on client/src/modules/journal/modals/search.modal.js:
```diff
@@ -52,42 +57,62 @@ function JournalSearchModalController(Instance, Projects, Notify,
   };
 
+  // Set displayLabel if the filters account is defined
+  if (filters.account_id) {
```
I still don't understand why this is even necessary.  Can'y we just do:
```js
displayValues.account_id = account.hrLabel;
```
and that be it?

All of these `if` statements are really ugly.  Let me fork this PR and see if I can get a cleaner version.

jniles reviewed (commented):
A few clean up things here.

All the `if (filters.key) {` statements look really ugly.  I'm going to try and fork this PR and make it cleaner.  My gut feels like it should be easier than this.

lomamech committed:
perf(Journal link Readable)
- use capital letter for name of controller
- remove unused property disable, name, required on
  ProjectSelectController

- Sanitaze projectSelect Component
- Remove all track of displayLabel for Each Filters
- use the display value for lookup in loggedchange

jniles commented on client/src/modules/journal/modals/search.modal.js:
```diff
@@ -117,16 +137,16 @@ function JournalSearchModalController(Instance, Projects, Notify,
 
     var loggedChanges = changes.getAll();
 
+    console.log('changes', loggedChanges);
```
This `console.log()` should be removed before merging.

jniles commented on client/src/js/components/bhProjectSelect.js:
```diff
@@ -0,0 +1,36 @@
+  };
+
+  // fires the onSelectCallback bound to the component boundary
+  $ctrl.onSelect = function ($item, $model) {
```
If you don't use the `$model`, you can leave it out.

jniles reviewed (commented):
Looks good to me.  There's a console log that should be removed and then this can land.  Great work!

lomamech committed:
refactor(Component): remove unused property in Service and Project

lomamech committed:
Resolve conflict

lomamech committed:
Fix conflict of Branch

jniles reviewed (approved):
Looks good to me!

bors r+
